### DISCUSSION TITLE
trace_raw question

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#79] [Q&A] [answered]
Hi @mjhoptics 

I have been puzzled by something - in the input to `trace_raw`, what is `pt0` meant to be. The doc says its the starting point in coords of first surface. But the first surface is the object and that is typically `1e10` (approx) distance away from the `z=0`, yet the input `pt0` has `z=0`. 

Regards

**Comment by mjhoptics** [comment, marked as answer]
Yes, this is by "design". Inside of the main loop in `trace_raw`, we start with the ray exiting the previous surface/interface. The ray data (intersection point and direction cosines) are in the previous surface's coordinate system. This gets transformed into the coordinate system of the next surface by applying `tfrm_from_before`. Now the ray data is in the coordinate system of the next surface, the one we'll try to intersect with. If you look at the coordinates of the ray point after transformation, you'll find it's z coord is -1e10.

If you have a curved object surface (e.g. typically how a dome projection system would be modeled), the z coord at the object for off-axis object points will be non-zero. This is what the intersection calculation before the main loop is meant to handle.

"infinite" object distances have been modeled by using extremely large distances since before there was an IEEE standard for floating point arithmetic, and the associated `inf` and `nan` values. I did a little work trying to make `inf` work as a legitimate distance, but if you don't do everything just right, you get a cascade of `nan's` in your results. Also, it's quite reasonable to take a lens designed for infinite conjugates and focus it for closer distances. This makes the code/calculations easier if you always assume the object distance is a finite (if large) floating point number.

**Reply by dibyendumajumdar** [reply]
> Yes, this is by "design". Inside of the main loop in `trace_raw`, we start with the ray exiting the previous surface/interface. The ray data (intersection point and direction cosines) are in the previous surface's coordinate system. This gets transformed into the coordinate system of the next surface by applying `tfrm_from_before`. Now the ray data is in the coordinate system of the next surface, the one we'll try to intersect with. If you look at the coordinates of the ray point after transformation, you'll find it's z coord is -1e10.
> 

Thank you for the explanation; Do I understand correctly then that the object plane is not in the coordinate system of the lens?

**Reply by mjhoptics** [reply]
I'm not sure I'd describe it like that. In the `seq_model`, there are 2 lists, `lcl_tfrms` and `gbl_tfrms`. The `gbl_tfrms` gives the position and orientation of all of the surfaces, including the object surface, with respect to the first surface (typically the first lens surface). If you look at the gbl_tfrm for the object surface, you'll see the position is `[0, 0, -1e10]`. So we have the information necessary to express the position and orientation of any surface with respect to any other surface.

You can call `seq_model.compute_global_coords` with any surface number and get the coordinates of all other surfaces with respect to that surface.

**Reply by dibyendumajumdar** [reply]
I see, okay thank you!
